### PR TITLE
Theme helper for search filters

### PR DESCRIPTION
**Pull request by gravityrail**
Some refactors and a theme helper for rendering current filters.

- [x] refactor helper functions to remove widget instance dependency
- [x] (unrelated) renames widget and related files from `*-search-filter-widget-*` to `*-search-widget-*` (since it is not just filters now)
- [x] create function to render filters on front-end

To test:

**Step 1: Add the following code to the `search.php` template in your theme:**

```php
<?php if ( class_exists('Jetpack_Search_Template_Tags' ) ): ?>
	<h2>Filter posts</h2>
	<?php Jetpack_Search_Template_Tags::render_available_filters(); ?>
<?php endif; ?>
```

**Step 2: Add some custom filters via a quick plugin:**

```php
<?php

/*
 * Plugin Name: Custom Woo search powered by Jetpack
 * Plugin URI: https://jetpack.com
 * Description: Search WooCommerce using Jetpack Search
 * Author: Automattic
 * Version: 0.01
 * Author URI: https://jetpack.com
 * License: GPL2+
 * Text Domain: woosearch
 * Domain Path: /languages/
 */

function woo_search_setup_filters() {
	if ( class_exists( 'Jetpack_Search' ) ) {
		Jetpack_Search::instance()->set_filters(array(
			'Month' => array(
				'type'     => 'date_histogram',
				'field'    => 'post_date',
				'interval' => 'month',
				'count'    => 10,
				'min_doc_count' => 1,
			),
			'Year' => array(
				'type'     => 'date_histogram',
				'field'    => 'post_date',
				'interval' => 'year',
				'count'    => 10,
				'min_doc_count' => 1,
			),
			'Categories' => array(
				'type'     => 'taxonomy',
				'taxonomy' => 'product_cat',
				'count'    => 10,
			),
		) );
	} else {
		// error_log("Jetpack search does not exist");
	}
}

add_action( 'init', 'woo_search_setup_filters' );
```

**Step 3: Load the search results page**

You should see something like this (I am not a designer ;) )

filter-posts-template-tag

### CODE REVIEW

**Comment by gibrown** [comment]
I'm not really sure what the design should be for this to be honest. Is there a user request for this?This seems like a huge change to try and get in when the beta is about to get released. Feels way more important to make sure that the release goes out and that we catch and fix bugs.

**Comment by gravityrail** [comment]
@gibrown this was requested by @kellychoffman as a critical piece for experienced site builders, since not all search result pages have a widget area, and this gives them some flexibility in how and where to render this functionality.

This PR also includes some refactors that I think make the code a bit more compact and easier to read, and has better separation of concerns between widget admin code and front-end rendering code, plus a couple of fixes for notices etc. that got triggered along the way.

File: modules/search/class.jetpack-search-template-tags.php [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+	public static function render_available_filters( $filters = null, $post_types = null ) {
+		if ( is_null( $filters ) ) {
+			$filters = Jetpack_Search::instance()->get_filters();
+		}
+
+		if ( is_null( $post_types ) ) {
+			$post_types = get_post_types( array( 'exclude_from_search' => false ) );
+		}
+
+		error_log(print_r($filters,1));
```
Comment (zinigor) [inline modules/search/class.jetpack-search-template-tags.php]: Debug logging detected :)

Comment (gravityrail) [reply modules/search/class.jetpack-search-template-tags.php]: Fixed - whoops!

**Comment by jeherve** [comment]
Would it be possible to edit the Search support doc to add such an example to the docs once 5.8 lands? I think it would be very useful.

**Pull request merged**